invoices/settlement_interceptor.go

```diff
@@ -0,0 +1,267 @@
+package invoices
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/lntypes"
+)
+
+// SafeCallback is a thread safe wrapper around an InterceptorClientCallback.
+type SafeCallback struct {
+	// mu is a mutex that protects the callback field.
+	mu sync.Mutex
+
+	// callback is the client callback function that is called when an
+	// invoice is intercepted. This function gives the client the ability to
+	// determine how the invoice should be settled.
+	callback InterceptorClientCallback
+}
+
+// Set sets the client callback function.
+func (sc *SafeCallback) Set(callback InterceptorClientCallback) {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+
+	sc.callback = callback
+}
+
+// Exec calls the client callback function.
+func (sc *SafeCallback) Exec(invoice Invoice) error {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+
+	return sc.callback(invoice)
+}
+
+// IsSet returns true if the client callback function is set.
+func (sc *SafeCallback) IsSet() bool {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+
+	return sc.callback != nil
+}
+
+type InvoiceResolution struct {
+	HtlcResolution    HtlcResolution
+	InvoiceUpdateDesc fn.Option[InvoiceUpdateDesc]
+}
+
+// InterceptSession creates a session that is returned to the caller when an
+// invoice is submitted to this service. This session allows the caller to block
+// until the invoice is processed.
+type InterceptSession struct {
+	ResolutionChan chan InvoiceResolution
+
+	// Quit is a channel that is closed when the session is no longer
+	// needed.
+	Quit chan struct{}
+}
+
+// SettlementInterceptor is a service that intercepts invoices during the
+// settlement phase, enabling a subscribed client to determine the settlement
+// outcome.
+type SettlementInterceptor struct {
+	wg   sync.WaitGroup
+	quit chan struct{}
+
+	// newInvoices is a channel queue of invoices that are waiting for
+	// client settlement decisions.
+	newInvoices chan Invoice
+
+	// callback is a client defined function that is called when an invoice
+	// is intercepted. This function gives the client the ability to
+	// determine the settlement outcome.
+	clientCallback SafeCallback
+
+	// interceptSessions is a map of intercept sessions that are used to
+	// store the client query/response session for a given invoice payment
+	// hash.
+	interceptSessions sync.Map
```

Comment:
Can have this be `lntuils.SyncMap`. Then the type of the key/value are explicit.